File: src/main/java/com/baidu/algorithm/dp/_121_Best_Time_to_Buy_and_Sell_Stock.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        if prices is None or len(prices) < 2:
            return 0

        left_min, max_profit = prices[0], 0
        for i in range(1, len(prices)):
            max_profit = max(max_profit, prices[i] - left_min)
            left_min = min(left_min, prices[i])

        return max_profit

File: src/main/java/com/baidu/algorithm/dp/_121_Best_Time_to_Buy_and_Sell_Stock.java
/*
 * Copyright (C) 2018 Baidu, Inc. All Rights Reserved.
 */
package com.baidu.algorithm.dp;

/**
 * _121_Best_Time_to_Buy_and_Sell_Stock
 *
 * @author xuhaoran01
 */
public class _121_Best_Time_to_Buy_and_Sell_Stock {

    public int maxProfit(int[] prices) {
        if(prices == null || prices.length == 0) {
            return 0;
        }

        int leftMin = prices[0], res = 0;
        for(int i = 1; i < prices.length; i++) {
            res = Math.max(res, prices[i] - leftMin);
            leftMin = Math.min(leftMin, prices[i]);
        }

        return res;
    }
}
